lib/semantics/resolve-labels.cc

```diff
@@ -892,18 +892,20 @@ void CheckLabelDoConstraints(const SourceStmtList &dos,
       errorHandler.Say(position,
           parser::MessageFormattedText{
               "label '%u' is not in scope"_en_US, SayLabel(label)});
-    } else if (!doTarget.labeledStmtClassificationSet.test(
-                   TargetStatementEnum::Do) &&
-        !doTarget.labeledStmtClassificationSet.test(
-            TargetStatementEnum::CompatibleDo)) {
-      errorHandler.Say(doTarget.parserCharBlock,
-          parser::MessageFormattedText{
-              "'%u' invalid DO terminal statement"_err_en_US, SayLabel(label)});
     } else if (!doTarget.labeledStmtClassificationSet.test(
                    TargetStatementEnum::Do)) {
-      errorHandler.Say(doTarget.parserCharBlock,
-          parser::MessageFormattedText{
-              "'%u' invalid DO terminal statement"_en_US, SayLabel(label)});
+      if (!doTarget.labeledStmtClassificationSet.test(
+              TargetStatementEnum::CompatibleDo)) {
+        errorHandler.Say(doTarget.parserCharBlock,
+            parser::MessageFormattedText{
+                "Only an END DO or CONTINUE must be used to terminate a labeled DO loop"_err_en_US,
+                SayLabel(label)});
+      } else {
+        errorHandler.Say(doTarget.parserCharBlock,
+            parser::MessageFormattedText{
+                "Only an END DO or CONTINUE should be used to terminate a labeled DO loop"_en_US,
+                SayLabel(label)});
```

Comment:
`SayLabel` was used in the original messages because the label number was included in the message. Now that it's not, `SayLabel(label)` is an unused argument.